endpoint: trigger k8s sync controller on identity update

jibi opened the pull request:
When an endpoint's identity is updated, Cilium does not sync immediately
the new state with k8s, but rather waits up to 10 seconds for the
sync-to-k8s-ciliumendpoint controller to run, meaning that the the new
identity can remain unannounced for up to 10 seconds.

This commit fixes this by explicitly triggering the k8s sync controller
whenever an endpoint's identity is updated.

Fixes https://github.com/cilium/cilium/issues/15097

jibi committed:
controller: allow to manually trigger it

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

gandro commented on pkg/controller/manager.go:
```diff
@@ -135,7 +136,7 @@ func (m *Manager) removeController(ctrl *Controller) {
 }
 
-func (m *Manager) lookup(name string) *Controller {
+func (m *Manager) Lookup(name string) *Controller {
```
No action required, just a thought:

One thing that I just noticed is that this export here is the first export that gives access to a `Controller` struct outside of the package. All other methods of the `Manager` do not allow direct access to the `Controller`. While it seems safe (i.e. it seems that `m.mutex` is not protecting it), I wonder if it might be cleaner and more future-proof to expose a `TriggerController(name string)` method instead (similar to `TerminationChannel`), instead of allowing a a direct lookup.

jibi replied:
Yeah, makes sense, that's where `(e *Endpoint) TriggerController(name string)` should actually live :+1:

gandro commented on pkg/endpoint/endpoint.go:
```diff
@@ -366,6 +370,11 @@ func (e *Endpoint) UpdateController(name string, params controller.ControllerPar
 
+// TriggerController triggers the endpoint's controller with the specified name.
+func (e *Endpoint) TriggerController(name string) {
+	e.controllers.Lookup(name).Trigger()
```
Since this is an exported function, we need to check that `Lookup(name) != nil`.

gandro commented on pkg/endpoint/endpoint.go:
```diff
@@ -366,6 +370,11 @@ func (e *Endpoint) UpdateController(name string, params controller.ControllerPar
 }
 
+// TriggerController triggers the endpoint's controller with the specified name.
+func (e *Endpoint) TriggerController(name string) {
```
nit: I think this function does not necessarily need to be exported in the current implementation. We could even consider inlining it at the call site.

gandro reviewed (commented):
Thanks a ton for taking care of this! Looks good overall. I have some minor nits (none of which need to be blocking)

christarazi commented on pkg/endpoint/endpoint.go:
```diff
@@ -2335,3 +2343,12 @@ func (e *Endpoint) setDefaultPolicyConfig() {
+	defer e.mutex.Unlock()
+
+	e.endpointSyncControllerName = name
+}
```
I have a suggestion that is more aesthetic than it is functional:

What if we instead move `controllerNameOf()` into the endpoint package rather than storing an extra string in the `Endpoint` struct? This way the ep syncher watcher can call `controllerNameOf()` and then when we manually trigger the controller, we call `controllerNameOf()` to get the name of the controller to trigger. Thus, we can avoid storing extra data into the struct.

jibi replied:
Yep, makes sense :+1: I was initially a bit hesitant as I felt we were mixing responsibilities (since this function is about a _controller_ name) but in the end it also relates to endpoints so I think it's fine to move it in the `endpoint` package

christarazi reviewed (commented):
LGTM. I have one suggestion below, but LMK what you think on it.

jibi committed:
endpoint: trigger k8s sync controller on identity update

When an endpoint's identity is updated, Cilium does not sync immediately
the new state with k8s, but rather waits up to 10 seconds for the
sync-to-k8s-ciliumendpoint controller to run, meaning that the the new
identity can remain unannounced for up to 10 seconds.

This commit fixes this by explicitly triggering the k8s sync controller
whenever an endpoint's identity is updated.

Fixes: #15097
Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
